File: frontend/src/views/app-views/students/add-student/validationRules.js
import { STUDENT_RULE } from './studentRules';

export const VALIDATED_FIELDS = [
  'code',
  'fullname',
  'dob',
  'classId',
  'email',
  'facebook',
  'username',
  'password',
  'hometown',
  'address',
  'hairColor',
];

export const getPasswordChecks = password => [
  { key: 'length', label: 'Tối thiểu 8 ký tự', passed: password.length >= 8 },
  { key: 'upper', label: 'Có chữ hoa', passed: /[A-Z]/.test(password) },
  { key: 'lower', label: 'Có chữ thường', passed: /[a-z]/.test(password) },
  { key: 'number', label: 'Có số', passed: /[0-9]/.test(password) },
  { key: 'special', label: 'Có ký tự đặc biệt', passed: /[^A-Za-z0-9\s]/.test(password) },
];

const getValue = (field, data, color) => {
  if (field === 'hairColor') {
    return color?.color;
  }

  return data[field];
};

const normalizeValue = value => (typeof value === 'string' ? value.trim() : value);

const hasDuplicate = (field, value, existingStudents = []) => {
  const normalizedValue = String(value).trim().toLowerCase();
  return existingStudents.some(
    student => String(student[field] || '').toLowerCase() === normalizedValue
  );
};

export const getFieldError = (field, data, color, existingStudents = []) => {
  const fieldRules = STUDENT_RULE.rules[field];
  const fieldMessages = STUDENT_RULE.message[field];

  if (!fieldRules || !fieldMessages) {
    return undefined;
  }

  const value = normalizeValue(getValue(field, data, color));

  if (fieldRules.required && !value) {
    return fieldMessages.required;
  }

  if (fieldRules.minLength && value && value.length < fieldRules.minLength) {
    return fieldMessages.minLength;
  }

  if (fieldRules.maxLength && value && value.length > fieldRules.maxLength) {
    return fieldMessages.maxLength;
  }

  if (fieldRules.beforeNow && value && new Date(value) >= new Date()) {
    return fieldMessages.beforeNow;
  }

  if (fieldRules.pattern && value && !fieldRules.pattern.test(value)) {
    return fieldMessages.pattern;
  }

  if (fieldRules.uniqueField && value && hasDuplicate(fieldRules.uniqueField, value, existingStudents)) {
    return fieldMessages.unique;
  }

  return undefined;
};

export const getPhotoError = file => {
  if (!file) {
    return undefined;
  }

  const photoRules = STUDENT_RULE.rules.photo;
  const photoMessages = STUDENT_RULE.message.photo;
  const extension = file.name.split('.').pop().toLowerCase();

  if (!photoRules.allowedTypes.includes(extension)) {
    return photoMessages.allowedTypes;
  }

  if (file.size > photoRules.maxSize) {
    return photoMessages.maxSize;
  }

  return undefined;
};

File: backend/services/validation_rules.py
CLASS_RULE = {
    "rules": {
        "code": {
            "required": True,
            "label": "Mã lớp",
            "unique": True,
        },
        "name": {
            "required": True,
            "label": "Tên lớp",
        },
    },
    "message": {
        "code": {
            "required": "Mã lớp là trường bắt buộc.",
            "unique": "Mã lớp đã tồn tại. Vui lòng nhập mã khác.",
        },
        "name": {
            "required": "Tên lớp là trường bắt buộc.",
        },
    },
}

STUDENT_RULE = {
    "rules": {
        "code": {
            "required": True,
            "label": "Mã học sinh",
            "unique": True,
        },
        "fullname": {
            "required": True,
            "label": "Họ và tên",
        },
        "dob": {
            "required": True,
            "label": "Ngày sinh",
        },
        "email": {
            "required": True,
            "required_on_copy": False,
            "label": "Email",
            "unique": True,
        },
        "class_id": {
            "required": True,
            "label": "Lớp học",
            "convert": True,
        },
        "username": {
            "required": True,
            "label": "Tài khoản",
            "unique": True,
        },
        "password": {
            "required": True,
            "label": "Mật khẩu",
        },
    },
    "message": {
        "code": {
            "required": "Mã học sinh là trường bắt buộc.",
            "unique": "Mã học sinh đã tồn tại. Vui lòng nhập mã khác.",
        },
        "fullname": {
            "required": "Họ và tên là trường bắt buộc.",
        },
        "dob": {
            "required": "Ngày sinh là trường bắt buộc.",
        },
        "email": {
            "required": "Email là trường bắt buộc.",
            "unique": "Email đã tồn tại. Vui lòng nhập email khác.",
        },
        "class_id": {
            "required": "Lớp học là trường bắt buộc.",
            "convert": "Lớp học không tồn tại. Vui lòng chọn lớp hợp lệ.",
        },
        "username": {
            "required": "Tài khoản là trường bắt buộc.",
            "unique": "Tài khoản đã tồn tại. Vui lòng nhập tài khoản khác.",
        },
        "password": {
            "required": "Mật khẩu là trường bắt buộc.",
        },
    },
}

MODEL_RULES = {
    "tra_class": CLASS_RULE,
    "tra_student": STUDENT_RULE,
}